Iter1

iudanet opened the pull request.

iudanet committed:
Initialize project structure and add core components

- Add docker-compose.yaml for PostgreSQL service - Create go.mod file -
Implement config package with parsing and tests - Add models for HTTP
requests/responses - Setup basic service and repository interfaces - Add
main.go placeholder with commented config setup

iudanet committed:
Add Makefile and test dependencies

The commit includes: - New Makefile with test and static analysis
targets - Added testify dependency in go.mod - New go.sum file with
dependency checksums - Refactored config tests to use testify assertions
- Renamed config fields to follow Go naming conventions

iudanet committed:
Restructure test targets and initialize config in main

iudanet committed:
Add PostgreSQL support with migrations and Makefile updates

- Add PostgreSQL repository implementation with migration support - Add
goose migration tool and initial user table migration - Update Makefile
with new targets for running, testing, and migrations - Add required
dependencies to go.mod - Initialize repo in main.go with database
connection

iudanet committed:
Add retry logic for database connections and migrations

- Implement retry mechanism with exponential backoff - Add PostgreSQL
error code detection for retriable errors - Improve error handling in
database initialization - Update Config.String() output format for
better readability - Add context support for database operations -
Replace panic with graceful exit on database connection failure

iudanet committed:
Optimize Postgres config and migration handling

- Tune Postgres performance with optimized parameters - Add context
support and status check to migrations - Define timeouts as constants
for ping and migrate operations - Tighten healthcheck intervals for
faster container readiness detection

iudanet committed:
Add database migration method to postgres repo

The migration logic is now separated from the repository initialization
and exposed through the Migrate method on the Repositories interface.

iudanet committed:
Remove deprecated Compose file version specification

The version field is no longer needed in modern Docker Compose files.
Also renamed models.go to repo.go for better clarity.

iudanet committed:
Add GolangCI-Lint configuration and integrate with Makefile

- Add .golangci.yml with all linters enabled - Update Makefile to
include lint targets and dependencies - Fix various linting issues in
code files - Add golangci-lint to prepare target

iudanet committed:
Implement HTTP server with handlers and graceful shutdown

- Add HTTP server with graceful shutdown handling - Create handlers for
auth, orders, balance and withdrawals - Add config flag for database
migration - Set up service layer with basic methods - Register all
routes in a new ServeMux

iudanet committed:
Refactor service constructor to use shorter name

iudanet committed:
Implement JWT authentication and user registration

- Add JWT dependency and crypto package - Create users_auth table
migration - Implement auth handlers with token generation - Add secret
key to config - Update service and repo layers for auth operations

Implement JWT authentication and user registration

- Add proper error handling and status codes for auth endpoints - Return
JWT token in Authorization header - Check for existing users during
registration - Update test configuration for auth testing

iudanet committed:
Implement order upload and retrieval functionality

- Add orders table migration - Implement order repository methods - Add
order service with Luhn validation - Update handlers for order upload
and retrieval - Fix typo in protected routes path - Rename test_ci_auth
to test_ci_one

iudanet committed:
Add order accrual processing system with worker and database changes

- Add worker service to periodically check and update order accruals -
Change accrual fields from int to float64 in models and database - Add
new database migration for accrual column type change - Implement
accrual client for external system communication - Add new Makefile
target for running accrual service - Extend repository interface with
order processing methods

iudanet committed:
Implement user balance functionality

- Add withdrawals table migration - Update BalanceResponse model to use
float64 for withdrawn - Implement balance calculation in repository -
Add balance endpoint handler with auth check - Expose balance methods
through service layer

iudanet committed:
mvp

iudanet committed:
Convert balance and withdrawals to use cents internally

iudanet committed:
Round balance values to 2 decimal places

- Add helper function for rounding currency values - Fix balance
calculation logic in users repo - Reduce worker poll interval to 500ms -
Add logging for withdrawal creation

iudanet committed:
Update Go version in CI workflows to 1.24

iudanet committed:
Add GitHub workflow for golangci-lint

iudanet committed:
fix version lint

alekstet commented on .golangci.yml:
```diff
@@ -0,0 +1,242 @@
+    - wsl_v5
+    - zerologlint
+  # Disable specific linters.
+  disable:
```
Круто что включил часть линтеров, а не ненужные отключил

alekstet commented on cmd/gophermart/main.go:
```diff
@@ -1,3 +1,75 @@
+		os.Exit(1)
+	}
+	if cfg.Migrate {
+		err = repo.Migrate(ctx)
```
Круто что прокатываешь миграции при старте приложения

alekstet commented on internal/handlers/auth.go:
```diff
@@ -0,0 +1,79 @@
+	err := s.svc.Register(r.Context(), req.Login, req.Password)
+	if err != nil {
+		var pqErr *pq.Error
+		if errors.As(err, &pqErr) && pqErr.Code == pgerrcode.UniqueViolation {
```
У тебя хендлер не должен знать что ты используешь базу постгре, это все должно быть скрыто за слоями абстракций

alekstet commented on internal/handlers/balanse.go:
```diff
@@ -0,0 +1,85 @@
+	}
+
+	// Конвертируем копейки в рубли с точностью до 2 знаков
+	currentRub := float64(currentCents) / 100
```
Это должно быть в слое сервиса

alekstet commented on internal/models/orderuser.go:
```diff
@@ -0,0 +1,25 @@
+// - `NEW` — заказ загружен в систему, но не попал в обработку;
+// - `PROCESSING` — вознаграждение за заказ рассчитывается;
+// - `INVALID` — система расчёта вознаграждений отказала в расчёте;
+// - `PROCESSED` — данные по заказу проверены и информация о расчёте успешно получена.
```
Круто что делаешь комментарии

alekstet commented on internal/repo/postgres/order_worker.go:
```diff
@@ -0,0 +1,107 @@
+) ([]models.OrderUser, error) {
+	rows, err := r.db.QueryContext(
+		ctx,
+		`SELECT number, status
```
Как вариант попробуй https://github.com/Masterminds/squirrel

iudanet committed:
Add mock generation and tests for handlers and service

The commit introduces several key changes: 1. Adds mock generation
targets to Makefile using mockgen 2. Implements comprehensive tests for
handlers (register, login, orders) 3. Adds tests for service layer (auth
and order functionality) 4. Includes utility tests for Luhn validation
5. Generates mock implementations for service and repository interfaces

The tests cover both happy paths and error cases, ensuring proper
validation and error handling throughout the application. Mock
generation enables isolated testing of components without requiring real
database connections.

alekstet commented on internal/repo/postgres/postgres.go:
```diff
@@ -0,0 +1,65 @@
+
+	// Используем retry для подключения к базе данных
+	err = retry.WithRetry(func() error {
+		db, err = sql.Open("postgres", dsn)
```
Можешь попробовать pgx или pgxpool

alekstet commented on internal/repo/postgres/users.go:
```diff
@@ -0,0 +1,159 @@
+func (r *postgresRepo) CreateUser(ctx context.Context, login, passwordHash string) error {
+	_, err := r.db.ExecContext(
+		ctx,
+		"INSERT INTO users_auth (login, password_hash) VALUES ($1, $2)",
```
Круто что используешь хэш пароля

alekstet commented on internal/repo/repo.go:
```diff
@@ -0,0 +1,30 @@
+	CreateOrder(ctx context.Context, userID int64, number string) error
+	GetOrderByNumber(ctx context.Context, number string) (*models.OrderUser, error)
+	GetOrdersByUserID(ctx context.Context, userID int64) ([]models.OrderUser, error)
+	GetOrderOwner(ctx context.Context, number string) (int64, error)
```
Как будто интерфейс слишком большой 
Может стоит подумать над разделением на части по доменам

iudanet committed:
Update test case names to use camelCase and fix accrual client behavior

- Change test case names from snake_case to camelCase across test files
- Modify accrual client to return error on 429 status instead of
retrying - Add new accrual client tests covering different response
scenarios

iudanet committed:
Standardize test case names to camelCase

iudanet committed:
Add Go test workflow and update Makefile test command

- Add GitHub Actions workflow for running Go tests on PRs and pushes -
Update Makefile test command to include verbose output and race detector

iudanet committed:
Update GitHub Actions upload-artifact from v3 to v4

iudanet committed:
Remove outdated users table migration

iudanet committed:
Replace PostgreSQL unique violation check with ErrUserAlreadyExists

iudanet committed:
Move balance rounding logic to service layer

The balance conversion from cents to rubles with rounding is now handled
in the service layer rather than the handler. This includes: - Moving
the rounding function to service/utils.go - Updating mock service
interfaces to return float64 - Removing redundant math import from
handler

iudanet committed:
Add mock implementations for repository interfaces

The commit adds mock implementations for UserRepository,
OrderRepository, BalanceRepository, and Migrator interfaces using
gomock. These mocks will be used for testing purposes.

iudanet committed:
Remove unused GetBalance method from UserService interface

iudanet committed:
Add mock for AccrualClientInterface and update worker to use interface

This commit: - Introduces AccrualClientInterface and implements it on
AccrualClient - Updates Worker to use the interface instead of concrete
type - Adds mock generation for the interface in Makefile - Includes
comprehensive tests for worker order processing scenarios

iudanet committed:
Refactor auth handling to use middleware and request context

- Move auth logic from utils.go to new middleware.go - Remove checkAuth
helper function - Update all handlers to get userID from request context
- Add auth middleware to protected routes - Add comprehensive tests for
auth middleware

iudanet committed:
Change OrderAccrualResponse.Accrual from pointer to value type

The field is now a non-pointer float64 since we always want to represent
accrual values, even when zero. Removed helper function ptrFloat64() as
it's no longer needed.

iudanet committed:
Refactor auth and luhn logic into separate packages

iudanet committed:
Add Swagger documentation support

The commit adds Swagger/OpenAPI documentation support to the project.
Key changes include:

1. Added swaggo/swag dependency for generating API docs 2. Created
Swagger annotations for all API endpoints in handler files 3. Added swag
init command to Makefile for generating docs 4. Implemented Swagger UI
endpoint setup 5. Added docs generation to test target in Makefile

The documentation provides interactive API documentation that can be
accessed via the /swagger/ endpoint. Each handler now includes detailed
Swagger annotations describing the endpoint's purpose, parameters,
responses, and security requirements.

The Makefile was updated to include swag_gen target and added it to the
test target dependencies to ensure docs are generated during testing.

s-shpak commented:
test
